c2/cuda/groth16_cuda.cu

```diff
@@ -525,24 +611,57 @@ RustError::by_value generate_groth16_proofs_c(const Assignment<fr_t> provers[],
 
                 {
                     msm_t<bucket_t, point_t, affine_t, scalar_t> msm{nullptr,
-                        tail_msm_l_bases.size()};
+                        std::max({split_vectors_l.tail_msm_scalars[0].size(),
+                                 split_vectors_a.tail_msm_scalars[0].size(),
+                                 split_vectors_b.tail_msm_scalars[0].size()})};
 
                     for (size_t c = circuit0; c < circuit0+num_circuits; c++) {
                         // tail MSM l - on GPU
+                        if (!l_skip_batch_add)
                         msm.invoke(results.l[c], tail_msm_l_bases,
                                    split_vectors_l.tail_msm_scalars[c], true);
+                        else
+                        msm.invoke(results.l[c], points_l.data(), points_l.size(),
+                                   split_vectors_l.tail_msm_scalars[c].data(), true);
+
                         if (caught_exception)
                             return;
 
                         // tail MSM a - on GPU
+                        if (!a_skip_batch_add)
                         msm.invoke(results.a[c], tail_msm_a_bases,
                                    split_vectors_a.tail_msm_scalars[c], true);
+                        else
+                        msm.invoke(results.a[c], points_a.data(), points_a.size(),
+                                   split_vectors_a.tail_msm_scalars[c].data(), true);
                         if (caught_exception)
                             return;
 
                         // tail MSM b_g1 - on GPU
+                        if (!b_skip_batch_add)
                         msm.invoke(results.b_g1[c], tail_msm_b_g1_bases,
                                    split_vectors_b.tail_msm_scalars[c], true);
+                        else
+                        msm.invoke(results.b_g1[c], points_b_g1.data(), points_b_g1.size(),
+                                   split_vectors_b.tail_msm_scalars[c].data(), true);
+                        if (caught_exception)
+                            return;
+                    }
+                }
+
+                if (G2_MSM_ON_GPU && !b_skip_batch_add)
+                {
+                    msm_t<bucket_fp2_t, point_fp2_t, affine_fp2_t, scalar_t>
+                        msm{nullptr, split_vectors_b.tail_msm_scalars[0].size()};
+                    for (size_t c = circuit0; c < circuit0+num_circuits; c++) {
+                        // tail MSM b_g2 - on GPU
+                        if (!b_skip_batch_add)
+                        msm.invoke(results.b_g2[c], tail_msm_b_g2_bases,
+                            split_vectors_b.tail_msm_scalars[c], true);
+                        else
+                        msm.invoke(results.b_g2[c], points_b_g2.data(), points_b_g2.size(),
+                            split_vectors_b.tail_msm_scalars[c].data(), true);
```

Comment:
I don't follow... For starters the inner if-else is not actually conditional, because the `!b_skip_batch_add` is always true on the line 658... Because of the line 652... Right? But then, wasn't the whole point to go for g2 msm on GPU specifically in the case we **skip** the batch addition? Which is `else` here...

Reply:
This seems like it will a case of balance between how fast the CPU and the GPU is compared to each other. Currently on a 3080, 32 GiB sector's window_post's G2 MSM takes 20~ seconds. The CPU is very fast on the same machine and overlapping other GPU operations with G2 MSM on the CPU perfectly is possible. It has me wondering if it makes sense to transfer more work from the GPU to the CPU. But it will be system-specific and also will depend on how many cores are actually available during C2.

Reply:
Yes, it's the about balancing. But that's not the issue I had. My understanding was that taking G2 MSM in window post to GPU is more or less the point of the exercise. Because G2 MSM in window post is so much heavier and CPU has no change to do in time. And I fail to see that it would be actually executed on GPU specifically in window post...

Reply:
The previous version made the G2 MSM execute on the CPU if it's a window_post.

> Because G2 MSM in window post is so much heavier and CPU has no change to do in time.

The CPU is able to complete G2 MSM while the GPU is busy doing NTTs and G1 MSMs. window_post with G2 MSM done on GPU takes 57~ seconds, but window_post with G2 MSM on the CPU takes only 37~ seconds. And in each of these cases, the GPU spends 37~ seconds on NTTs and G1 MSMs.

Reply:
Ah! In other words I've got the picture up side down:-):-):-) It sounds like there is **never** a reason to execute G2 MSM on GPU, be it tail or full one. At least not on the reference system that is. It might be beneficial if the CPU thread pool is throttled, right? But one can always make it user's responsibility to do configure the system correctly... Either way, one can wonder if it's possible to do some "job stealing". I mean if CPU has completed G2 MSM, and GPU is sufficiently far from the corresponding G1 MSM, one can "steal" it and do on CPU.

Reply:
> one can wonder if it's possible to do some "job stealing".

And conversely, in case when we generate multiple proofs, if GPU is done with G1 MSMs, and the throttled thread pool still struggles with G2 tail MSMs, GPU can "steal" one or two...

Reply:
This got me wondering too. I suppose one can do some tricks with CUDA streams but it will be complex, especially here. I'll dig through some documentation.

Reply:
There are gpu.sync-s... I mean those might be sufficient...

Reply:
> There are gpu.sync-s... I mean those might be sufficient...

But let's leave it for later...